xapian-letor/ranker/listnet_ranker.cc

```diff
@@ -0,0 +1,264 @@
+			gradient[k] += second_term;
+
+		}
+	}
+
+	return gradient;
+}
+
+static void
+updateParameters(vector<double> & new_parameters, vector<double> gradient, double learning_rate) {
+	int num = new_parameters.size();
+
+	for (int i = 0; i < num; i++){
+		new_parameters[i] -= gradient[i] * learning_rate;
+	}
+
+}
+
+static void
+batchLearning(const std::vector<FeatureVector> & fvv, vector<double> & new_parameters, double learning_rate) {
+
+	// initialize Probability distributions of y and z
+	vector< vector<double> > prob = initializeProbability(fvv, new_parameters);
+	// compute gradient
+	vector<double> gradient = calculateGradient(fvv, prob);
+	// update parameters: w = w - gradient * learningRate
+	updateParameters(new_parameters, gradient, learning_rate);
+
+}
+
+void
+ListNETRanker::train_model() {
+
+	std::cout << "ListNET model begin to train..." << endl;
+
+	// get the training data
+	std::vector<Xapian::FeatureVector> fvv = get_traindata();
+	int fvv_len = fvv.size();
+
+	int feature_cnt = -1;
+	if (fvv_len != 0) {
+		feature_cnt = fvv[0].get_fcount();
+	}
+	else {
+		std::cout << "The training data is NULL!" << endl;
+		exit(1); // TODO: Throw exception
+	}
+
+	// initialize the parameters for neural network
+	vector<double> new_parameters(feature_cnt, 0.0);
+
+	// iterations
+	for (int iter_num = 1; iter_num < iterations; ++iter_num) {
+
+		batchLearning(fvv, new_parameters, learning_rate);
+	}
+
+    parameters = new_parameters;
+
+}
+
+void
+ListNETRanker::save_model_to_file(const char* output_filename) {
+
+    ofstream parameters_file;
+    parameters_file.open(output_filename);
+
+    int parameters_size = parameters.size();
+
+    std::ostringstream oss;
+    oss.precision (std::numeric_limits<double>::digits10);
+
+    for (int i = 0; i < parameters_size; ++i) {
+        oss << parameters[i] << endl;
+    }
+
+    parameters_file << oss.str();
+    parameters_file.close();
+
+    cout<< "ListNET model saved to \"" << output_filename << "\"" << endl;
+}
+
+void
+ListNETRanker::load_model_from_file(const char* model_filename) {
+
+	vector<double> loaded_parameters;
+
+    fstream train_file (model_filename, ios::in);
+    if (!train_file.good()) {
+        cout << "No ListNET parameters file found" << endl;
+        exit(1); // TODO: should throw an exception
+    }
+
+    while (train_file.peek() != EOF) {
+
+        double parameter;
+        train_file >> parameter;
+        loaded_parameters.push_back(parameter);
+    }
+
+    train_file.close();
+
+    loaded_parameters.pop_back();
+    parameters = loaded_parameters;
+
+    cout<< "ListNET model loaded from: \"" << model_filename << "\""  << endl;
+}
+
+std::vector<FeatureVector>
+ListNETRanker::rank(std::vector<FeatureVector> & fvv) {
+
+    std::vector<FeatureVector> testfvv = fvv;
+    int testfvvsize = testfvv.size();
+
+    cout << "Performing ranking using ListNET model..." << endl;
+
+    std::vector<double> new_parameters = parameters;
```

Comment:
You aren't modifying `parameters`, so is there a reason not to just use it directly?